### PR TITLE
[Core] Robust date validation

### PR DESCRIPTION
**Pull request by alisterdev**
- [x] Make sure that a date of y format or my/ym format gets converted to
yyyy-mm-dd accepted by database. Furthermore check if a date is valid
(i.e 123123-01-01 is not valid)
- [x] Updates to jquery date picker used as a fallback in Firefox, Safari and other non `<input type='date'/>` friendly browsers.

>**Bugfix for 17.0.2**

### CODE REVIEW

File: php/libraries/LorisForm.class.inc [hunk 1]
```diff
@@ -1956,31 +1956,35 @@ class LorisForm
...
+                if ((strtolower($el['options']['format']) === 'my') ||
+                    (strtolower($el['options']['format']) === 'ym')) {
+                    $date = sprintf("%s-01", $value);
+                }
+                // Make sure the date is valid
+                $dt = DateTime::createFromFormat('Y-m-d', $date);
+                $isValidDate = $dt && $dt->format('Y-m-d') == $date;
+
+                if (!$isValidDate) {
+                    throw new Exception("Please input a valid date!");
```
Comment (alisterdev) [inline php/libraries/LorisForm.class.inc]: throw LorisException here

**Comment by alisterdev** [comment]
Travis fails because I disabled manual input for date fields in Firefox (to force users to actually click on jquery datepicker and prevent input of invalid dates)

In theory, clicking on `$dates[0]` should bring the calendar UI and clicking on `$day[0]` should fill the input with first day of the month
```php
$dates = $this->webDriver->findElements(WebDriverBy::cssSelector(".input-date"));
$days = $this->webDriver->findElements(WebDriverBy::cssSelector(".ui-datepicker-calendar td[data-handler='selectDay']"))
```

**Comment by kongtiaowang** [comment]
@alisterdev 
https://github.com/aces/Loris/pull/2574 
I cloned your PR and modified the automation test files in this PR.

**Comment by alisterdev** [comment]
[Fix test for broken date](https://trello.com/c/by3kW0r2/85-fix-test-for-broken-date)

**Comment by alisterdev** [comment]
@kongtiaowang Thanks for help!

**Comment by MounaSafiHarab** [comment]
@ridz1208 
can you please test this when you are back to see if it resolved the dates issue on Firefox?

**Comment by ridz1208** [comment]
@alisterdev 

 - tests with YYYY-MM-DD format passed
   - on chrome dates start at 1970, on firefox and safari they start from 2017 (only issue)

File: php/libraries/LorisForm.class.inc [hunk 2]
```diff
@@ -1956,31 +1956,36 @@ class LorisForm
...
+                    || (strtolower($el['options']['format']) === 'ym')
+                ) {
+                    $date = sprintf("%s-01", $value);
+                }
+                // Make sure the date is valid
+                $dt          = DateTime::createFromFormat('Y-m-d', $date);
+                $isValidDate = $dt && $dt->format('Y-m-d') == $date;
+
+                if (!$isValidDate) {
+                    throw new LorisException("Please input a valid date!");
```
Comment (ridz1208) [inline php/libraries/LorisForm.class.inc]: @alisterdev  this should be an error on the same page not an exception I think

Comment (alisterdev) [inline php/libraries/LorisForm.class.inc]: The problem is that the date validation was not done in `validate()` function which actually assigns the errors, but in a function that runs after error checking (which I updated for more robust checks. 

I'm open to take all the date related code and move it to validate, but I wanted @jstirling91 or @driusan to weight in as they are more familiar with the code.

**Comment by alisterdev** [comment]
@ridz1208 Replaced LorisException with an error

**Comment by ridz1208** [comment]
@alisterdev 
 - managed to get February 275760 passed it
 - only get `Please input a valid date in 'Confirm Date of Birth'!` when dates dont match
 - not able to get Please input a valid date in 'Date of Birth'!`

**Comment by alisterdev** [comment]
@ridz1208 Please test it one last time

File: php/libraries/LorisForm.class.inc [hunk 3]
```diff
@@ -1956,31 +1971,39 @@ class LorisForm
...
+                // Make sure the date is valid
+                $dt          = DateTime::createFromFormat('Y-m-d', $date);
+                $isValidDate = $dt && $dt->format('Y-m-d') == $date;
+
+                if (!$isValidDate) {
+                    $this->setElementError(
+                        $name,
+                        "Please input a valid date in '" . $el['label'] . "'!"
+                    );
+                }
```
Comment (jstirling91) [inline php/libraries/LorisForm.class.inc]: this should move to the validate function doing something like
```
if($el['type'] === 'date') {
    $date            = ($values[$elName]) ? $values[$elName] : $this->getValue($el['name']);
    $dt                = DateTime::createFromFormat('Y-m-d', $date);
    $isValidDate = $dt && $dt->format('Y-m-d') == $date;
    if (!$isValidDate) {
        $this->setElementError(
            $name,
             "Please input a valid date in '" . $el['label'] . "'!"
          );
          $errors[$el['name']] = "date fail";
      }
}
```

**Comment by alisterdev** [comment]
@jstirling91 Updated

**Comment by jstirling91** [comment]
This looks good to me, though we might want to make this a little more advanced for 17.1. With the current implementation, if a date is within a group it wont be checked for validity. We should make a rule that behaves similar to numeric for future iteration.

**Comment by alisterdev** [comment]
@ridz1208 If you think this is good, tag Dave to merge it

**Comment by ridz1208** [comment]
@driusan All good, please merge

**Comment by alisterdev** [comment]
@driusan +1

**Pull request merged**